design/auth.md

```diff
@@ -0,0 +1,270 @@
+
+```json
+{
+  ....
+  "scope": [
+    "build:3001": "write"
+  ]
+}
+```
+
+_Note, in this example `projects` and `jobs` are not listed because it's assumed they have read permission based on the `build` listed._
+
+## Signing Changes
+
+Signing keys should be rotated periodically.  For us to do that, we should have an identity for each public/private key as well as an expiration.  That gives the clients the information to expire keys appropriately and be able to pull new keys when they are made available.
+
+```json
+{
+  "WbXr5q1JbSx8JDzTBm6u": {
+    "key": "MIIBIjANBgkqhkiG9w0BAQEFA.....",
+    "exp": 1486465521
+  }
+}
+```
+
+## JWT Content Changes
+
+*Standard Fields from JWT Spec:*
+
+- **`iss` (Issuer)** The issuer of the token, typically the fqdn of the authorization server.
+- **`sub` (Subject)** The subject of the token; the name or id of the client which requested it. This should be empty (`""`) if the client did not authenticate.
+- **`aud` (Audience)** The intended audience of the token; the name or id of the service which will verify the token to authorize the client/subject.
+- **`exp` (Expiration)** The token should only be considered valid up to this specified date and time.
+- **`nbf` (Not Before)** The token should not be considered valid before this specified date and time.
+- **`iat` (Issued At)** Specifies the date and time which the Authorization server generated this token.
+- **`jti` (JWT ID)** A unique identifier for this token. Can be used by the intended audience to prevent replays of the token.
+
+*Screwdriver specific fields:*
+
+- **`pki` (Public Key ID)** An identifier of the public key published by the Auth server to verify the signature of this JWT.
+- **`profile`** Information to appropriately display who the current authenticated user is.
+
+### User Refresh Token
+
+```json
+{
+  "iss": "auth.screwdriver.cd",
+  "sub": "user:stjohnjohnson",
+  "aud": "auth.screwdriver.cd",
+  "iat": 1486421770,
+  "nbf": 1486421770,
+  "exp": 1486464970,
+  "jti": "TnpZeE5EYzFZbU16WXpFM1ltTTJPVFppTVdJNVl6SsI5ZyTTBNbVF3WXpVd1",
+  "pki": "WbXr5q1JbSx8JDzTBm6u",
+  "profile": {
+    "name": "St. John Johnson",
+    "url": "https://github.com/stjohnjohnson",
+    "img": "https://avatars2.githubusercontent.com/u/622065?v=3&s=460"
+  }
+}
+```
+
+### User Bearer Token
+
+```json
+{
+  "iss": "auth.screwdriver.cd",
+  "sub": "user:stjohnjohnson",
+  "aud": "auth.screwdriver.cd",
+  "iat": 1486421770,
+  "nbf": 1486421770,
+  "exp": 1486421970,
+  "jti": "TnpZeE5EYzFZbU16WXpFM1ltTTJPVFppTVdJNVl6SsI5ZyTTBNbVF3WXpVd1",
+  "pki": "WbXr5q1JbSx8JDzTBm6u",
+  "profile": {
+    "name": "St. John Johnson",
+    "url": "https://github.com/stjohnjohnson",
+    "img": "https://avatars2.githubusercontent.com/u/622065?v=3&s=460"
+  },
+  "scope": {
+    "project:20": "read",
+    "job:100": "write",
+    "job:101": "write",
+    "job:102": "write",
+    "job:103": "write"
+  }
+}
+```
+
+### Build Refresh Token
```

Comment:
We may want to talk about when refreshing a token is allowed. For example, if a build has a status that indicates it has been completed, any refresh request for that build should be denied.